src/index.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * ⚠️ WARNING! This is just an implementation of the resolver spec
+ * for demo purposes. It is subject to change, or possibly fall out
+ * of sync with the specification. In case of a deviation, prefer
+ * the official specification over this example library.
+ */
+import { mergeTokenSets } from './set.js';
+import type { DTCGTokens, Resolver } from './types.js';
+import { mergeTokens } from './utils.js';
+
+export function createResolver<T extends Record<string, any> = DTCGTokens>(
+  tokenMap: Record<string, T>,
+  resolver: Resolver,
+) {
+  if (!Object.keys(tokenMap ?? {}).length) {
+    throw new Error(`Empty token map! No tokens to resolve`);
+  }
+
+  const tokens: T = resolver?.sets?.length
+    ? mergeTokenSets(
+        resolver.sets
+          .map(({ name, values }) => {
+            if (!values?.length) {
+              throw new Error(`Token set ${name} can’t contain empty array of values`);
+            }
+            return values.map<T>((id) => getTokens(id));
+          })
+          .flat(),
+      )
+    : ({} as T);
+
+  function getTokens(id: keyof typeof tokenMap) {
+    if (!(id in tokenMap)) {
+      throw new Error(`Tokens "${id}" missing in tokenMap!`);
+    }
+    return tokenMap[id];
+  }
+
+  return {
+    tokens,
+    getTokens,
+    apply(values: Record<string, string>): T {
```

Comment:
I don’t really feel strongly about the API, but in general I was going for the API of:

```ts
const r = createResolver(tokens, resolverJSON);
r.apply({ foo: "bar", bar: "baz", … });
```

This way you can keep a resolver around for a while, and “apply” values over and over again without losing your already-flattened sets.

I didn’t put a ton of thought into it; happy to come up with an alternate API